src/swt.rs

```diff
@@ -0,0 +1,106 @@
+// Copyright 2019 King's College London.
+// Created by the Software Development Team <http://soft-dev.org/>.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use super::{MirTrace, ThreadTracer};
+/// Software tracing via ykrustc.
+use core::yk_swt::{self, MirLoc};
+use libc;
+use std::ops::Drop;
+
+/// A trace collected via software tracing.
+/// Since the trace is a heap-allocated C buffer, we represent it as a pointer and a length.
+struct SWTMirTrace {
+    buf: *mut MirLoc,
+    len: usize
+}
+
+impl MirTrace for SWTMirTrace {
+    fn len(&self) -> usize {
+        self.len
+    }
+
+    fn loc(&self, idx: usize) -> &MirLoc {
+        if idx > isize::max_value() as usize {
+            panic!("index too large for ptr arithmetic");
+        }
+
+        if idx >= self.len {
+            panic!("out of bounds index");
+        }
+
+        unsafe { &*self.buf.offset(idx as isize) }
+    }
+}
+
+impl Drop for SWTMirTrace {
+    fn drop(&mut self) {
+        unsafe { libc::free(self.buf as *mut libc::c_void) };
+    }
+}
+
+/// Softare thread tracer.
+struct SWTThreadTracer;
+
+impl ThreadTracer for SWTThreadTracer {
+    fn stop_tracing(self: Box<Self>) -> Option<Box<dyn MirTrace>> {
+        match yk_swt::stop_tracing() {
```

Comment:
Maybe `yk_swt::stop_tracing().map(|(buf, len)| Some(Box::new(...))`?

Reply:
Ah yes. I never was good at all of these combinators :)